### PR TITLE
Adding HostedService that checks for new BTCPayServer version on GitHub once a day

### PR DESCRIPTION
**Pull request by rockstardev**
Resolves: #1420 

Policy to check for new version needs to enabled (false by default)

image

The check is disabled when developing (`NetworkType == NetworkType.Regtest && Environment.IsDevelopment()`).

I've implemented it as part of Policy settings since to me setting this over there seemed more intuitive... if you prefer to rather have it as part of `BTCPayServerOptions`, let me know.

### CODE REVIEW

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        protected async Task LoopVersionCheck()
+        {
+            await ProcessVersionCheck();
+            await Task.Delay(TimeSpan.FromDays(1), Cancellation);
+        }
+
+        public async Task ProcessVersionCheck()
+        {
+            var policies = await _settingsRepository.GetSettingAsync<PoliciesSettings>() ?? new PoliciesSettings();
+            if (policies.CheckForNewVersions && !_env.IsDeveloping)
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:42]: remove `!_env.IsDeveloping`.

`CheckForNewVersions` should be false by default, I prefer not pushing people to update.

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        Task<string> Fetch(CancellationToken cancellation);
+    }
+
+    public class GithubVersionFetcher : IVersionFetcher
+    {
+        private readonly HttpClient _httpClient;
+        public GithubVersionFetcher()
+        {
+            _httpClient = new HttpClient();
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:75]: Use the IHttpClietnFactory.

Comment (rockstardev) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs]: d280617

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs (hunk 1 above)
Comment (rockstardev) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs]: d280617

**Comment by pavlenex** [comment]
Shouldn't we have User preferences for notifications? I get this particular one makes sense in Server settings, but then they'd be scattered around?

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        private readonly HttpClient _httpClient;
+        public GithubVersionFetcher(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient(nameof(GithubVersionFetcher));
+            _httpClient.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json");
+            _httpClient.DefaultRequestHeaders.Add("User-Agent", "BTCPayServer/NewVersionChecker");
+        }
+
+        public void Dispose()
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:80]: I don't think you call that anywhere. That said, you don't have to dispose HttpClient... this is one of those weird type that defy conventional wisdom...

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            _httpClient.Dispose();
+        }
+
+        public async Task<string> Fetch(CancellationToken cancellation)
+        {
+            const string url = "https://api.github.com/repos/btcpayserver/btcpayserver/releases/latest";
+            var resp = await _httpClient.GetAsync(url, cancellation);
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:88]: though I think you need to dispose the resp here!

**Comment by NicolasDorier** [comment]
Later we will just do a user level way to filter notifications. That is orthogonal to this PR.

**Comment by NicolasDorier** [comment]
@rockstardev outside my two nit comments, you can merge this.

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs (hunk 3 above)
Comment (Kukks) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:80]: Yeah this is just a magical call that .net invokes itself somehow.

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        Task<string> Fetch(CancellationToken cancellation);
+    }
+
+    public class GithubVersionFetcher : IVersionFetcher, IDisposable
+    {
+        private readonly HttpClient _httpClient;
+        public GithubVersionFetcher(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient(nameof(GithubVersionFetcher));
```
Comment (Kukks) [inline BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:75]: @NicolasDorier  I was wondering if maybe we should use the tor http client if it is available to do this call to not ping a third party constantly? (Also think it should be an option on rates so maybe it should be a separate PR)

**Comment by NicolasDorier** [comment]
I agree with @Kukks on `BTCPAY_UPSTREAM_GITHUB_REPO`

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs (hunk 5 above)
Comment (NicolasDorier) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:75]: I don't think we should aim at using TOR for protecting the privacy of the server itself, this is too hard, untestable, and will be quite unreliable.

This is better handled by using a VPN or putting the whole network behind TOR.

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            await ProcessVersionCheck();
+            await Task.Delay(TimeSpan.FromDays(1), Cancellation);
+        }
+
+        public async Task ProcessVersionCheck()
+        {
+            var policies = await _settingsRepository.GetSettingAsync<PoliciesSettings>() ?? new PoliciesSettings();
+            if (policies.CheckForNewVersions)
+            {
+                var tag = await _versionFetcher.Fetch(Cancellation);
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:44]: what if that throw? (ie, github unavailable)

This will probably execute `ProcessVersionCheck` in a tight loop and spamming the logs.

**Comment by NicolasDorier** [comment]
I think https://github.com/btcpayserver/btcpayserver/pull/1788#discussion_r463445424 is important as well

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs (hunk 5 above)
Comment (Kukks) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:75]: Yeah fair

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs (hunk 6 above)
Comment (Kukks) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:44]: Yeah I guess a try catch with either waiting another day or giving it a 1hour delay?

Comment (rockstardev) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs:55]: 11c09ea

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs (hunk 4 above)
Comment (rockstardev) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs]: 11c09ea

File: BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs (hunk 3 above)
Comment (rockstardev) [reply BTCPayServer/HostedServices/NewVersionCheckerHostedService.cs]: You two are ganging up on these code reviews of my PRs so much, that now there is a duplication of threads. 

image

Resolved: 11c09ea , same as below.

**Comment by rockstardev** [comment]
> Can you remove the public async Task<IActionResult> Generate(string version) in NotificationsController now?

No, it's used in test of notification in UnitTest1.cs

> One concern I have is not being able to specify the url of the repo to ping for versions of forks. Personally, would add a new config option "BTCPAY_UPSTREAM_GITHUB_REPO" and set it default to https://github.com/btcpayserver/btcpayserver

I've added option to directly set path to JSON file so that there isn't any magic string appending. Thus if someone wants to serve their update JSON file somewhere other than GitHub - they can, it just needs to be in format `{ "tag_name": "v1.2.3.4" }`

> Another concern is what happens if we do a feature test release like we had for the vlndseedbackup thing? Would this grab it and announce it as a new version?

I've decided to use same regex we use for circleci

For user notifications preferences - I spoke with @pavlenex and will be doing page to allow each user to select which notifications it wants to receive. When that's done I'll initialize each admins update check default from BTCPayServerOptions (instead of server policy).

File: BTCPayServer/Configuration/DefaultConfiguration.cs [hunk 7]
```diff
@@ -38,6 +38,8 @@ protected override CommandLineApplication CreateCommandLineApplicationCore()
             app.Option("--sshtrustedfingerprints", "SSH Host public key fingerprint or sha256 (default: empty, it will allow untrusted connections)", CommandOptionType.SingleValue);
             app.Option("--torrcfile", "Path to torrc file containing hidden services directories (default: empty)", CommandOptionType.SingleValue);
             app.Option("--socksendpoint", "Socks endpoint to connect to onion urls (default: empty)", CommandOptionType.SingleValue);
+            app.Option("--updatecheck", $"Enabling once a day check for new releases (default: false)", CommandOptionType.SingleValue);
+            app.Option("--updateurl", $"Url location used for updatecheck (default: https://api.github.com/repos/btcpayserver/btcpayserver/releases/latest)", CommandOptionType.SingleValue);
```
Comment (NicolasDorier) [inline BTCPayServer/Configuration/DefaultConfiguration.cs:42]: can you make it only in one variable? If not specified or empty, then the check is disabled.

Comment (rockstardev) [reply BTCPayServer/Configuration/DefaultConfiguration.cs:42]: Ah, this is what I was considering @NicolasDorier... but what happens when person has update url other than default and doesn't want to configure it each time.

This would also require our users to copy-paste ` https://api.github.com/repos/btcpayserver/btcpayserver/releases/latest` every time they want to enable it. What would be the way we're going to deploy this with `btcpayserver-docker`? BTCPAY_UPDATEURL empty and then users required to populate it to enable?

Comment (NicolasDorier) [reply BTCPayServer/Configuration/DefaultConfiguration.cs:42]: The default should be empty. It is btcpayserver-docker job to pass the right URL.

Comment (rockstardev) [reply BTCPayServer/Configuration/DefaultConfiguration.cs]: 16eedf4
